### ISSUE TITLE
Swap out fiber implementation

### ISSUE THREAD

**Issue by rsmmr** [zeek/spicy#561]
The current fiber/co-routine implementation borrows low-level code from libtask for context switching that uses a fixed stack size per fiber. Two problems there:

1. We can't catch if a fiber runs out of stack space. We'll just crash in ways that probably open doors widely for all kinds of stack-smashing attacks.

2. Not being able to grow the stack dynamically means we need to estimate upfront how much stack fibers need max, and then always allocate the worst-case, which is very wasteful.

I've been looking around for alternatives to libtask's code. It's difficult to find something that's both (1) addressing those issues, (2) well maintained, (3) cross-platform (for what we need), and  (4) narrow enough to not pull in a bunch of additional infrastructure we don't need (e.g., a full-featured event scheduler).

The best candidate I've found so is https://github.com/hnes/libaco. I'd like to give that a try. It can use a shared stack for all fibers. That comes with a performance impact (because it copies stack frames around) but I can't tell how much impact that would have for us. It also protects its stack with guard pages against overflow to at least crash reliably.

**Comment by bbannier** [comment]
It looks like libaco [does not support macos at this point](https://github.com/hnes/libaco/blob/fe379c1a5964783aef26685e1167312aa40e146b/acosw.S#L3-L4). Using different fiber implementations for different host platforms seems like something we wouldn't want to do, so I do not think that libaco works for us.

**Comment by bbannier** [comment]
> It looks like libaco [does not support macos at this point](https://github.com/hnes/libaco/blob/fe379c1a5964783aef26685e1167312aa40e146b/acosw.S#L3-L4). Using different fiber implementations for different host platforms seems like something we wouldn't want to do, so I do not think that libaco works for us.

I looked more carefully and it when build correctly libaco actually does support macos. Since the including build setup broke with the `grep` on my macos, I checked this with the following cmake setup:
```.cmake
project(aco C ASM)

add_library(aco OBJECT acosw.S aco.c)

set(THREADS_PREFER_PTHREAD_FLAG ON)
find_package(Threads REQUIRED)

enable_testing()

set(tests "test_aco_tutorial_0;test_aco_tutorial_1;test_aco_tutorial_2;test_aco_tutorial_3;test_aco_tutorial_4;test_aco_tutorial_5;test_aco_tutorial_6;test_aco_synopsis;test_aco_benchmark")
foreach (test ${tests})
    add_executable(${test} ${test}.c)
    target_link_libraries(${test} PRIVATE aco Threads::Threads)
    add_test(${test} ${test})
endforeach (test)
```

I'll see how we can consume it in hilti-rt.

**Comment by rsmmr** [comment]
Sounds good. I'll still make a summary of fiber/co-routine implementation out there, too.

**Comment by bbannier** [comment]
I looked into using [libcoro](http://software.schmorp.de/pkg/libcoro.html) and it looks like it currently only handles Windows and Linux, see [here](http://cvs.schmorp.de/libcoro/coro.h?revision=1.60&view=markup#l321).

**Comment by rsmmr** [comment]
I think it was this comment that gave me hope: `

```
-DCORO_ASM
--
149 | *
150 | *    Hand coded assembly, known to work only on a few architectures/ABI:
151 | *    GCC + arm7/x86/IA32/amd64/x86_64 + GNU/Linux and a few BSDs. Fastest
152 | *    choice, if it works.
```

**Comment by bbannier** [comment]
Closed by #601.
